File: pyref/atlasopt/result.py
"""ソルバー共通の結果型。"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class SolveResult:
    """最適化の結果。

    points: 消費ポイント数(= 選択ノード数 - 1。root/スタートノードはコスト0)
    nodes: 選択ノード集合(root含む)
    status: "optimal"(最適性証明付き) / "feasible"(実行可能解のみ) / "infeasible"
    dual_bound: ソルバーの下界。status == "optimal" なら points に一致する
    ignored_terminals: 除外設定などでrootから到達不能となり無視されたterminal
    solve_time: ILP求解時間(秒)。前処理は含まない
    """

    points: int
    nodes: frozenset[str]
    status: str
    dual_bound: float = 0.0
    ignored_terminals: tuple[str, ...] = ()
    solve_time: float = 0.0

File: web/src/solver/result.ts
/** 求解結果(pyref/atlasopt/result.py の移植)。 */

export type SolveStatus = "optimal" | "feasible" | "infeasible";

export interface SolveResult {
  /** 消費ポイント数 = |nodes| - 1(rootはコスト0)。infeasible 時は -1 */
  points: number;
  /** 選択ノード集合(root含む) */
  nodes: ReadonlySet<string>;
  status: SolveStatus;
  /** 除外/到達不能で無視された terminal(辞書順) */
  ignoredTerminals: readonly string[];
  /** ILP求解時間(秒)。前処理は含まない */
  solveTime: number;
  /**
   * ソルバーの下界。pyref は HiGHS の mip_dual_bound を常に持つが、highs-js は
   * これを公開しないため optimal 時のみ points と同値を設定する(意図的な差分)。
   * 「最適性の証明」判定は status === "optimal" に一本化されている。
   */
  dualBound?: number;
}
